### PR TITLE
Fixed expires

### PR DESCRIPTION
**Pull request by Ragazzo**
Follow up https://github.com/zendframework/zf2/pull/7557 

[RFC](http://tools.ietf.org/html/rfc6265#section-4.1)

### CODE REVIEW

**Comment by Martin-P** [comment]
Just as a reminder: please add unittests.

**Comment by Ragazzo** [comment]
Afaik, empty expires means that cookie is only for current user session after which it should be deleted. Need to review if this fix is correct according this statetment.

File: test/Header/SetCookieTest.php [hunk 1]
```diff
@@ -343,6 +343,12 @@ public function testDoesNotAcceptCookieNameFromArbitraryLocationInHeaderValue()
         $this->assertNotEquals('leo_auth_token', $setCookieHeader->getName());
     }
 
+    public function testEmptyExpiresSection()
+    {
+        $cookie = 'Set-Cookie:path=/; Expires=; HttpOnly;';
+        $setCookieHeader = SetCookie::fromString($cookie);
```
Comment (Martin-P) [inline test/Header/SetCookieTest.php]: Where is the actual test if the `expires` is `null`? You are now testing nothing.

Comment (Ragazzo) [inline test/Header/SetCookieTest.php]: i test condition if expires is empty string, remove fix to code and exception will be thrown

Comment (Ragazzo) [inline test/Header/SetCookieTest.php]: or you are referring to missing `->isNull` on `getExpires` ?

Comment (Martin-P) [inline test/Header/SetCookieTest.php]: There is no assertion, so nothing is tested. You can use a try/catch block to test if an exception is thrown. Also please update the method name so it reflects what you are actually testing.

You can do it like this:
```php
public function testEmptyExpiresDoesNotThrowException()
{
    $cookieString = 'Set-Cookie:path=/; Expires=; HttpOnly;';

    try {
        SetCookie::fromString($cookie);
    } catch (\Exception $exception) {
        $this->fail('Empty "expires" value should not throw an exception');
    }
    
    $this->pass('No exception was thrown for empty "expires" value');
}
```

Comment (Ragazzo) [inline test/Header/SetCookieTest.php]: dont need to add try catch, you can use `@expectedException`, overall i will only update code with `assertIsNull`, your fix is not correct. Tests should not always have assertions, if test has passed without exceptions than it is ok, pass or fail

Comment (Martin-P) [inline test/Header/SetCookieTest.php]: > you can use `@expectedException`

If you expect an exception you can use that, but in this case you expecting the opposite: no exception. You can not use annotations for that, because there is none for "not expecting an exception".

> Tests should not always have assertions, if test has passed without exceptions than it is ok, pass or fail

I agree it is not mandatory, but IMO it is better to be explicit when testing.

Comment (Maks3w) [inline test/Header/SetCookieTest.php]: Please assert `getExpires` returns the correct value.

Comment (Ragazzo) [inline test/Header/SetCookieTest.php]: Ok

Comment (Martin-P) [inline test/Header/SetCookieTest.php]: > Please assert `getExpires` returns the correct value.

I think that is a test on its own. One test to see there is no exception thrown and another test to see if `getExpires` returns `null`.

**Comment by Ragazzo** [comment]
@Maks3w done, ready for review

File: src/Header/SetCookie.php [hunk 2]
```diff
@@ -378,7 +378,7 @@ public function getMaxAge()
      */
     public function setExpires($expires)
     {
-        if ($expires === null) {
+        if ($expires === null || $expires === '') {
```
Comment (Maks3w) [inline src/Header/SetCookie.php]: Should `0` to be handle as null too?

Comment (Martin-P) [inline src/Header/SetCookie.php]: No, it seems `0` is not allowed. See: https://github.com/zendframework/zf2/issues/4684#issuecomment-19951416

**Comment by Martin-P** [comment]
@Ragazzo Can you provide a link to the specifications where empty `expires` is mentioned?

File: src/Header/SetCookie.php (hunk 2 above)
Comment (Maks3w) [inline src/Header/SetCookie.php]: @Martin-P that thread is about Expires **HEADER**

> If set to 0, or omitted, the cookie will expire at the end of the session (when the browser closes).
http://php.net/manual/en/function.setcookie.php

Comment (Ragazzo) [inline src/Header/SetCookie.php]: not sure, i've read rfc, but it is still unclear for me about this situation http://tools.ietf.org/html/rfc6265#section-4.1.2.1

Comment (Martin-P) [inline src/Header/SetCookie.php]: @Maks3w I know, but a cookie is just a header. Also the class name confirms this: `Zend\Http\`**`Header`**`\SetCookie`.

Comment (Maks3w) [inline src/Header/SetCookie.php]: @Martin-P Expires header and Set-Cookie headers are defined by different rules.

It's wrong assume a constraint for Proxy header affects to Authorization header. The same is applicable for Expires and Set-Cookie

**Comment by Ragazzo** [comment]
@Martin-P do let me know if other things need to be done, as for rfc as i said i am not sure about correct behavior in this case, but throwing exception on valid header attribute does not sound right too

**Comment by weierophinney** [comment]
@Ragazzo — per @Martin-P , can you provide the link to the `expires` specification in the Cookie header RFC so (a) others coming to the issue know where to look, and (b) we can validate the behavior is correct per the RFC?

**Comment by Ragazzo** [comment]
@weierophinney updated description, yet as i said it is still unclean for me about this situation in RFC

**Comment by ezimuel** [comment]
I close this PR due to inactivity.
